### PR TITLE
NED

### PR DESCRIPTION
**Pull request by Mehrad0711**
This rather small PR (as indicated by branch name) brings named entity disambiguation into genienlp.
I'm doing some final testing to make sure everything works but also want to get your early feedback on it.

### CODE REVIEW

**Comment by gcampax** [comment]
Yep, really small pull request. Can't really review this in detail. Is this all that is needed for bootleg?

I see *a lot* of hacks. Any reference to specific domains or hardcoded types will have to go (although we can merge this specific PR first). What is the plan for that?

Also, does this work at inference time? How? Have you tested it?

**Comment by s-jse** [comment]
I took a quick look at this. Where did all this Bootleg code come from? Can we just run Bootleg somewhere else (in its own repository, or at most a script here that does that) and get entities of the dataset in text format. Then everything else is normal `genienlp`?

**Comment by Mehrad0711** [comment]
> I see _a lot_ of hacks. Any reference to specific domains or hardcoded types will have to go (although we can merge this specific PR first). What is the plan for that?
- We should decide what to do with domain-specific type mappings. 1) We can remove them but that would hurt performance. 2) We can continue including more domains 3) We can provide a set of all thingtalk types for all domains and filter invalid bootleg types based on that. I prefer 3) as we get the best of both worlds.

> Also, does this work at inference time? How? Have you tested it?
- Yeah. In server mode, a bootleg annotator is loaded on gpu and extracts features for each batch of examples on the fly. I also separated bootleg from the task, so we can put annotator on a server and fire examples at it (instead of loading it for each validation).

> Can we just run Bootleg somewhere else (in its own repository, or at most a script here that does that)
- This is exactly what run_bootleg.py script does by leveraging Bootleg class. This class takes care of extracting mentions, running disambiguation, collecting features, and extracting types.

**Comment by s-jse** [comment]
My complaint was that this PR makes a lot of changes to the implementation of model embeddings, which were used in your experiments, but if the simpler "entities in text form" approach matched or outperformed those embedding changes, then we should just merge the simpler approach.

**Comment by gcampax** [comment]
> > I see _a lot_ of hacks. Any reference to specific domains or hardcoded types will have to go (although we can merge this specific PR first). What is the plan for that?
> 
>     * We should decide what to do with domain-specific type mappings. 1) We can remove them but that would hurt performance. 2) We can continue including more domains 3) We can provide a set of all thingtalk types for all domains and filter invalid bootleg types based on that. I prefer 3) as we get the best of both worlds.

4) We can let skill developers provide the appropriate type mapping by designing a file format that Genie can generate off of Thingpedia information.

> 
> > Also, does this work at inference time? How? Have you tested it?
> 
>     * Yeah. In server mode, a bootleg annotator is loaded on gpu and extracts features for each batch of examples on the fly. I also separated bootleg from the task, so we can put annotator on a server and fire examples at it (instead of loading it for each validation).

There is no GPU in the actual NLP servers. What is the expected design here, is bootleg running in the same container as the semantic parser or no? What about the database? Do you need the Bootleg JSONs in the server container?
Are there new arguments to `genienlp server` or new environment variables (it's hard to tell from the PR)?

> My complaint was that this PR makes a lot of changes to the implementation of model embeddings, which were used in your experiments, but if the simpler "entities in text form" approach matched or outperformed those embedding changes, then we should just merge the simpler approach.

Agreed actually. It would make reviewing easier, and avoid a lot of tricky hard to test code.

**Comment by Mehrad0711** [comment]
>but if the simpler "entities in text form" approach matched or outperformed those embedding changes, then we should just merge the simpler approach.

That's true for TransformerSeq2Seq models but for TransformerLSTM the text-based and non text-based approach performs similarly (on the spotify dataset I've experimented with). Besides we will need to report ablation results using different methods so I prefer if the working code is there. I have removed the non text-based code for TransformerSeq2Seq models.
But we can remove it once we start merging the code in paraphrase dir (including transformers_utils.py where the current models that use entity embedding are at) after this.

**Comment by Mehrad0711** [comment]
>There is no GPU in the actual NLP servers. What is the expected design here, is bootleg running in the same container as the semantic parser or no? What about the database? Do you need the Bootleg JSONs in the server container?
Are there new arguments to genienlp server or new environment variables (it's hard to tell from the PR)?

In server mode, it runs in the same container but it can be separate. I think it should be possible to run the annotator on CPU too with lower batch size. I will test it.
And yeah bootleg model and database files are loaded from disk. (Model is ~600MB and combined database files is ~3GB)
You would need to pass `--database_dir` in genienlp (or GENIENLP_DATABASE_DIR env variable when running `genie evaluate-server`) which indicates where the files are located at.

**Comment by gcampax** [comment]
> > There is no GPU in the actual NLP servers. What is the expected design here, is bootleg running in the same container as the semantic parser or no? What about the database? Do you need the Bootleg JSONs in the server container?
> > Are there new arguments to genienlp server or new environment variables (it's hard to tell from the PR)?
> 
> In server mode, it runs in the same container but it can be separate. I think it should be possible to run the annotator on CPU too with lower batch size. I will test it.

Lower batch size? Surely on CPU we can have higher batch size, not lower?

> And yeah bootleg model and database files are loaded from disk. (Model is ~600MB and combined database files is ~3GB)
> You would need to pass `--database_dir` in genienlp (or GENIENLP_DATABASE_DIR env variable when running `genie evaluate-server`) which indicates where the files are located at.

Is @jgd5 aware of these new requirements for running models?

**Comment by Mehrad0711** [comment]
> Lower batch size? Surely on CPU we can have higher batch size, not lower?

Sorry I meant higher...

So if you remember, we merged some changes for serve in genie-toolkit before to run oracle models (including passing answer to genienlp server, passing --database_dir flags, etc. in predictor.ts). I have made some minor changes (renaming mainly!) which I can PR later.

**Comment by gcampax** [comment]
> So if you remember, we merged some changes for serve in genie-toolkit before to run oracle models (including passing answer to genienlp server, passing --database_dir flags, etc. in preditor.ts). I have made some minor changes (renaming mainly!) which I can PR later.

Yes but Jim and I recently changed that code, so that at actual inference time (not during model evaluation, during actual deployment) genienlp is not spawned by genie-toolkit, it's run in a separate k8s container. So you need to coordinate with him if there are new arguments to add to the container.

File: genienlp/arguments.py
Comment (gcampax) [inline genienlp/arguments.py]: Please add "ner" or "ned" prefix to all the ner/ned specific arguments (at least those that don't have "bootleg" or "entity" in the name). The name is really confusing otherwise.

File: genienlp/data_utils/bootleg.py
Comment (gcampax) [inline genienlp/data_utils/bootleg.py:66]: When you train a multidomain Mario model, what domains do you pass?
This is ok for now because the PR needs to land, but you need to refactor this ASAP. Make some JSON file that Genie can generate with the information you need.

File: genienlp/data_utils/bootleg.py
Comment (gcampax) [inline genienlp/data_utils/bootleg.py:93]: Why is `type` a QID and not a ThingTalk type?

Why do you not traverse the wikidata graph up until you hit the right type instead of writing a bunch of if statements?

File: genienlp/data_utils/bootleg.py
Comment (gcampax) [inline genienlp/data_utils/bootleg.py:193]: This is just "human". It is not right in general. How did you choose these representative QIDs?

File: genienlp/data_utils/bootleg.py
Comment (gcampax) [inline genienlp/data_utils/bootleg.py:258]: What is the content of these three files? Are they just dictionaries?
If so, you can use the old [HashTable code](https://github.com/stanford-oval/genienlp/blob/69c0d5f801bf924a6a617edffcc6ca2c77971d66/genienlp/data_utils/hash_table.py) to create a binary on-disk representation of the dictionary that you can memory map. That will yield 0 memory usage, because the kernel will page in and out of disk as needed, and should solve our memory issues.

File: genienlp/data_utils/database_utils.py
Comment (gcampax) [inline genienlp/data_utils/database_utils.py:39]: Why do you need this two-level mapping instead of mapping directly to ThingTalk types in the earlier function?
In any case, this cannot be hardcoded.

File: genienlp/data_utils/database.py
Comment (gcampax) [inline genienlp/data_utils/database.py]: Can you load this lazily, so we don't load it unless it's used?

File: genienlp/data_utils/bootleg.py
Comment (gcampax) [inline genienlp/data_utils/bootleg.py:31]: Missing copyright notice (here and in other new files).

File: genienlp/data_utils/example.py
Comment (gcampax) [inline genienlp/data_utils/example.py]: Time to fix this TODO? pad_id needs to match the real pad_id presumably?

File: genienlp/data_utils/iterator.py
Comment (gcampax) [inline genienlp/data_utils/iterator.py]: Please restore the fact that the warning is only emitted once.

File: genienlp/data_utils/numericalizer.py
Comment (gcampax) [inline genienlp/data_utils/numericalizer.py:186]: Have you experimented using regular tokens instead of special tokens as entity boundaries, given pretraining?

File: genienlp/data_utils/remote_database.py
Comment (gcampax) [inline genienlp/data_utils/remote_database.py]: How is the config passed? Is it environment variables or arguments? Given it contains secrets, it should be either a config file or environment variables.

File: genienlp/models/transformer_lstm.py
Comment (gcampax) [inline genienlp/models/transformer_lstm.py:69]: Should use `isinstance` here.

File: genienlp/models/transformer_lstm.py
Comment (gcampax) [inline genienlp/models/transformer_lstm.py]: This code needs to handle save_directory the same way as the other branch (ie, not call from_pretrained if save_directory is set)

File: genienlp/server.py
Comment (gcampax) [inline genienlp/server.py]: Wat?
If you need the task to tell you which field is the sentence, you ask the task to tell you. You don't hide this logic in a random place.
Please refactor.

File: genienlp/server.py
Comment (gcampax) [inline genienlp/server.py:219]: The database needs to be stored in the same folder as the model, because there is no way to specify a different path to the server.

File: genienlp/tasks/almond_task.py
Comment (gcampax) [inline genienlp/tasks/almond_task.py]: ???
Please don't alter copyright notices, that's literally illegal.

File: genienlp/tasks/almond_task.py
Comment (gcampax) [inline genienlp/tasks/almond_task.py]: As noted by the highlighting, names that start with capital letters are reserved for classes. This is not a class.

File: genienlp/tasks/almond_task.py
Comment (gcampax) [inline genienlp/tasks/almond_task.py:84]: The Marisa Trie library is designed to load/write binary files directly from disk - that's how it becomes memory efficient. Please refactor.

File: genienlp/tasks/almond_utils.py
Comment (gcampax) [inline genienlp/tasks/almond_utils.py]: What's this function doing? The name is too generic.

**Review by gcampax — CHANGES_REQUESTED** [review]
There are a lot of small changes that I would like fixed before merging this, like the argument names. Then we should merge this, because the PR is already barely reviewable, and then you need to start working immediately on the major changes, because in this state I cannot use Bootleg.

File: genienlp/data_utils/bootleg.py
Comment (Mehrad0711) [reply genienlp/data_utils/bootleg.py:66]: Only Spotify. Yeah, domain-specific codes will be gone (or at least reduced) in the next PR.

File: genienlp/data_utils/bootleg.py
Comment (Mehrad0711) [reply genienlp/data_utils/bootleg.py:93]: That was the older implementation which was slower and had lower results for Spotify...

Comment (gcampax) [reply genienlp/data_utils/bootleg.py:93]: Which part was the older implementation, traversing the graph? Did you investigate why?

File: genienlp/data_utils/bootleg.py
Comment (Mehrad0711) [reply genienlp/data_utils/bootleg.py:193]: As long as you are training single domain (or multi-domain like Mario but you only want NED for one domain) the representative QID shouldn't matter. However, I put a TODO for multidomain (e.g. SQA multidomain) where all types should be granular so film actor --> Q10798782 , film director -->  Q2526255, etc. instead of all mapping to Q5.

Comment (gcampax) [reply genienlp/data_utils/bootleg.py:193]: Well, the representative QID matters when you extract the Bootleg type embedding right? Otherwise you wouldn't need a QID at all.

File: genienlp/data_utils/bootleg.py
Comment (Mehrad0711) [reply genienlp/data_utils/bootleg.py:93]: Yeah traversing. The dictionary that maps Qids to types is incomplete. Bootleg uses some heuristics such as comparing the frequency of a term appearing in Wikipedia to how often it appears as a labeled mention. If it appears as a mention less than <~3% of the time, they filter it out. Thus traversing doesn't always reach the type we want.

File: genienlp/data_utils/bootleg.py
Comment (Mehrad0711) [reply genienlp/data_utils/bootleg.py:258]: Interesting! I'll look into it.

File: genienlp/data_utils/bootleg.py
Comment (gcampax) [reply genienlp/data_utils/bootleg.py:93]: But we have the real Wikidata in Elastic Search. We can do a round of processing on the Bootleg list of types and map them to a ThingTalk type ahead of time with some ES searches. In fact, we only need to do this once and store the processed types, as long as the mapping of "root QID" to ThingTalk type doesn't change.

File: genienlp/data_utils/database_utils.py
Comment (Mehrad0711) [reply genienlp/data_utils/database_utils.py:39]: Mainly for transparency. I wanted to keep everything in Q-space and only map to TT types in the final stage.
Also, in the previous embedding-based approach, bootleg_post_process_types is optional and we can just look up type Qids in type2id mapping and embed them.

File: genienlp/data_utils/bootleg.py
Comment (Mehrad0711) [reply genienlp/data_utils/bootleg.py:93]: Yeah, we can use our ES to complete Bootleg databases. Will add it to the next PR's bucket list.

File: genienlp/data_utils/bootleg.py
Comment (Mehrad0711) [reply genienlp/data_utils/bootleg.py:193]: Yeah, but we are not using bootleg embeddings anymore cause they didn't help. In fact, I removed all the code for integration level 2. 
(P.S. We can switch to "dump_preds" mode vs "dump_embeds" mode which saves tons of space!)

File: genienlp/tasks/almond_task.py
Comment (Mehrad0711) [reply genienlp/tasks/almond_task.py]: I didn't change it! It was just copied partially. Fixed now...

File: genienlp/data_utils/numericalizer.py
Comment (Mehrad0711) [reply genienlp/data_utils/numericalizer.py:186]: I suspect it would make a difference cause adding types to text does not generate more `correct_without_param` errors but I can try.

File: genienlp/data_utils/remote_database.py
Comment (Mehrad0711) [reply genienlp/data_utils/remote_database.py]: It's a config file and should be located in `database_dir/es_material/elastic_config.json`

File: genienlp/models/transformer_lstm.py
Comment (Mehrad0711) [reply genienlp/models/transformer_lstm.py:69]: Well I was but then realized`isinstance` checks for inheritance so `isinstance(AutoConfig.from_pretrained('xlm-roberta-base'), BertConfig)` is True!
I can reorder ifs to check for XLMRobertaConfig first, but using `type` is cleaner?

File: genienlp/server.py
Comment (Mehrad0711) [reply genienlp/server.py:219]: But the database is independent of the model. Aren't we passing an ENV variable for the database?

File: genienlp/tasks/almond_task.py
Comment (Mehrad0711) [reply genienlp/tasks/almond_task.py:84]: So canonical2type is about 4G. I can create another json containing only canonical2type keys or I can save `canonical2type.keys()` to memory on the fly and load with marisa_trie. In the first case, we lose disk memory and the second approach might be slower than the current implementation? What do you think?

File: genienlp/server.py
Comment (Mehrad0711) [reply genienlp/server.py]: Sorry, not sure what you're suggesting here. `is_contextual()` returns either True or False based on the task.
Are you suggesting it should be a class attribute rather than a method?

**Comment by s-jse** [comment]
Can we finalize the genienlp version 0.6 release before merging this?

**Comment by Mehrad0711** [comment]
> Can we finalize the genienlp version 0.6 release before merging this?

If you want to merge #74 before this I'm ok with it since there's not a lot of overlapping changes. Is that PR ready?

**Comment by gcampax** [comment]
> Can we finalize the genienlp version 0.6 release before merging this?

As per the [release schedule](https://wiki.almond.stanford.edu/en/release-planning/two-point-oh), genienlp 0.6 will be released at the same time as Almond 2.0.
We can make more alpha releases of genienlp in meantime, if necessary. Is it necessary for any reason?

**Comment by s-jse** [comment]
> 
> 
> > Can we finalize the genienlp version 0.6 release before merging this?
> 
> As per the [release schedule](https://wiki.almond.stanford.edu/en/release-planning/two-point-oh), genienlp 0.6 will be released at the same time as Almond 2.0.
> We can make more alpha releases of genienlp in meantime, if necessary. Is it necessary for any reason?

I thought a release without all the bells and whistles of bootleg would be useful for other people who may want to use genienlp for simple NLP experiments/papers.

**Comment by gcampax** [comment]
Yes, we have that, it's genienlp 0.5 and lower.

File: genienlp/tasks/almond_task.py
Comment (Mehrad0711) [reply genienlp/tasks/almond_task.py:84]: Flagged for next PR.

**Review by gcampax — APPROVED** [review]
Yeah let's merge this and let's keep going.

**Pull request merged**
